### PR TITLE
Implement translations for claims messages

### PR DESCRIPTION
**Pull request by dan-jones-iw**
stole types from MiM

### CODE REVIEW

File: examples/secondary-care/homecare/acute/nominated-pharmacy/clinical-practitioner/4-Convert-Response-Claim-200_OK.xml [hunk 1]
```diff
@@ -41,6 +41,9 @@
     <subject>
       <DispenseClaimInformation classCode="INFO" moodCode="EVN">
         <id root="A2C05EC6-4292-4568-8200-BFAC8439DA04"/>
+        <code code="163541000000107" codeSystem="2.16.840.1.113883.2.1.3.2.4.15" displayName="Dispensed Medication - FocusActOrEvent (administrative concept)"/>
+        <effectiveTime value="PLACEHOLDER"/>
```
Comment (booshibooshi) [inline examples/secondary-care/homecare/acute/nominated-pharmacy/clinical-practitioner/4-Convert-Response-Claim-200_OK.xml:45]: should this placeholder be being replaced?

Comment (dan-jones-iw) [reply examples/secondary-care/homecare/acute/nominated-pharmacy/clinical-practitioner/4-Convert-Response-Claim-200_OK.xml:45]: yep, will sort

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  messageId: string,
+  timestamp: hl7V3.Timestamp,
+  logger: pino.Logger
+) {
+  //TODO - work out what this means and whether we're doing it:
+  // "Note: this must refer to the last one in the series if more
+  // than one dispense event was required to fulfil the prescription."
+  const supplyHeader = new hl7V3.DispenseClaimSupplyHeader(new hl7V3.GlobalIdentifier(messageId))
+
+  //TODO - repeat dispensing
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-claim.ts:108]: look in claim.item.extension (same url as dispensing)

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+}
+
+export function createSuppliedLineItem(
+  claim: fhir.Claim,
+  item: fhir.ClaimItem,
+  detail: fhir.ClaimItemDetail
+): hl7V3.DispenseClaimSuppliedLineItem {
+  const claimSequenceIdentifierExtension = getExtensionForUrl(
+    detail.extension,
+    "https://fhir.nhs.uk/StructureDefinition/Extension-ClaimSequenceIdentifier",
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-claim.ts:151]: this identifier might have to reference the line item id

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    "Claim.item.detail.extension"
+  ) as fhir.IdentifierExtension
+  const suppliedLineItem = new hl7V3.DispenseClaimSuppliedLineItem(
+    new hl7V3.GlobalIdentifier(claimSequenceIdentifierExtension.valueIdentifier.value)
+  )
+  suppliedLineItem.effectiveTime = hl7V3.Null.NOT_APPLICABLE
+  //TODO - repeat dispensing
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-claim.ts:158]: above

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+  suppliedLineItem.component = detail.subDetail.map(subDetail => {
+    const hl7SuppliedLineItemQuantity = createSuppliedLineItemQuantity(claim, item, detail, subDetail)
+    return new hl7V3.SuppliedLineItemComponent(hl7SuppliedLineItemQuantity)
+  })
+
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-claim.ts:163]: TODO pertinentInfo 1

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+  const statusReasonExtension = getExtensionForUrlOrNull(
+    detail.extension,
+    "https://fhir.nhs.uk/StructureDefinition/Extension-EPS-TaskBusinessStatusReason",
+    "Claim.item.extension"
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-claim.ts:167]: 165

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    pertinentInformation1,
+    pertinentInformation2
+  )
+}
+
+function getChargePaid(subDetail: fhir.ClaimItemSubDetail) {
+  const prescriptionChargeCoding = getCodeableConceptCodingForSystem(
+    subDetail.programCode,
+    "https://fhir.nhs.uk/CodeSystem/DM-prescription-charge",
+    "Claim.item.detail.programCode"
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-claim.ts:243]: subDetail

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts [hunk 8]
```diff
@@ -2,21 +2,30 @@ import {fhir, hl7V3} from "@models"
...
-import {dispenseNotificationIsRepeatDispensing} from "../index"
+import {
+  Consumable,
+  DispenseNotificationPertinentInformation2,
+  DispenseNotificationSuppliedLineItemQuantityPertinentInformation1,
+  RequestedManufacturedProduct,
+  SuppliedLineItemComponent1,
+  SupplyInstructions,
+  SupplyRequest
+} from "../../../../../../models/hl7-v3"
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-notification.ts:28]: oof

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts [hunk 9]
```diff
@@ -168,24 +246,24 @@ function createPertinentInformation1LineItem(
     hl7PertinentSuppliedLineItem.repeatNumber = getRepeatNumberFromRepeatInfoExtension(repeatInfo)
   }
 
-  return new hl7V3.DispenseNotificationPertinentInformation1LineItem(hl7PertinentSuppliedLineItem)
+  return new hl7V3.SupplyHeaderPertinentInformation1(hl7PertinentSuppliedLineItem)
 }
 
-function getRepeatNumberFromRepeatInfoExtension(
+export function getRepeatNumberFromRepeatInfoExtension(
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-notification.ts:252]: might not need export (double check)

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts [hunk 10]
```diff
@@ -168,24 +246,24 @@ function createPertinentInformation1LineItem(
...
   repeatInfoExtension: fhir.ExtensionExtension<fhir.IntegerExtension>
 ): hl7V3.Interval<hl7V3.NumericValue> {
   const numberOfRepeatsIssuedExtension = getExtensionForUrl(
     repeatInfoExtension.extension,
     "numberOfRepeatsIssued",
     /* eslint-disable-next-line max-len */
-    'MedicationDispense.extension("https://fhir.nhs.uk/StructureDefinition/Extension-EPS-RepeatInformation").extension'
+    "MedicationDispense.extension(\"https://fhir.nhs.uk/StructureDefinition/Extension-EPS-RepeatInformation\").extension"
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-notification.ts:259]: undo change?

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts [hunk 11]
```diff
@@ -195,29 +273,22 @@ function getRepeatNumberFromRepeatInfoExtension(
   )
 }
 
-function createSuppliedLineItemQuantity(
-  hl7SuppliedLineItemQuantitySnomedCode: hl7V3.SnomedCode,
+export function createSuppliedLineItemQuantity(
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-notification.ts:276]: might not need exporting

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts [hunk 12]
```diff
@@ -264,15 +316,18 @@ export function getPrescriptionStatus(fhirFirstMedicationDispense: fhir.Medicati
...
+export function getPrescriptionItemId(fhirMedicationDispense: fhir.MedicationDispense): string {
   return getIdentifierValueForSystem(
     fhirMedicationDispense.authorizingPrescription.map(e => e.identifier),
     "https://fhir.nhs.uk/Id/prescription-order-item-number",
     "MedicationDispense.authorizingPrescription.identifier"
   )
 }
 
-function getDosageInstruction(fhirMedicationDispense: fhir.MedicationDispense, logger: pino.Logger) {
+export function getDosageInstruction(
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-notification.ts:327]: 319, 327, 346

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts [hunk 13]
```diff
@@ -315,92 +366,39 @@ function createPertinentPrescriptionId(fhirFirstMedicationDispense: fhir.Medicat
   return new hl7V3.PrescriptionId(hl7PertinentPrescriptionId)
 }
 
-function createPriorOriginalRef(firstMedicationDispense: fhir.MedicationDispense) {
+export function createOriginalPrescriptionRef(
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/request/dispense/dispense-notification.ts:369]: 354, 369

File: examples/secondary-care/homecare/acute/nominated-pharmacy/clinical-practitioner/4-Process-Request-Claim-200_OK.json [hunk 14]
```diff
@@ -44,28 +44,26 @@
...
       }
     },
     {
-      "fullUrl": "urn:uuid:4509B70D-D8B8-EA03-1105-64557CB54A29",
+      "fullUrl": "urn:uuid:1a5e542f-47f9-4358-96b4-f9780ca46ac3",
       "resource": {
         "resourceType": "Claim",
-        "identifier": [
+        "id": "e8e75676-9d30-45f8-8b09-696818e7f4ca",
+        "identifier":  [
```
Comment (dan-jones-iw) [inline examples/secondary-care/homecare/acute/nominated-pharmacy/clinical-practitioner/4-Process-Request-Claim-200_OK.json:58]: ctrl alt l

File: models/fhir/common.ts [hunk 15]
```diff
@@ -1,4 +1,7 @@
 import {LosslessNumber} from "lossless-json"
+import { PersonOrOrganization } from "."
+import { fhir } from ".."
```
Comment (dan-jones-iw) [inline models/fhir/common.ts:3]: wut

File: models/fhir/common.ts [hunk 16]
```diff
@@ -51,6 +54,7 @@ export function createCoding(system: string, code: string, display: string): Cod
   }
 }
 
+// Matt and Dan -
```
Comment (dan-jones-iw) [inline models/fhir/common.ts:57]: ?

File: models/fhir/extension.ts [hunk 17]
```diff
@@ -25,6 +25,10 @@ export interface ReferenceExtension<T extends common.Resource> extends Extension
   valueReference: common.Reference<T>
 }
 
+export interface IdentifierReferenceExtension<T extends common.Resource> extends Extension {
```
Comment (dan-jones-iw) [inline models/fhir/extension.ts:28]: double check

File: models/hl7-v3/dispense-common.ts [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+import * as codes from "./codes"
+import {ElementCompact} from "xml-js"
+import * as prescription from "./prescription"
+import * as lineItem from "./line-item"
+
+/*
+ * An act relationship that associates the Dispense focal act with
+ * SupplyHeader - the primary act of the PSIS clinical message.
+ */
+export class DispenseCommonPertinentInformation1<T extends SupplyHeader<DispenseCommonSuppliedLineItem>>
```
Comment (dan-jones-iw) [inline models/hl7-v3/dispense-common.ts:11]: increase duplication :+1:

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts (hunk 3 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-claim.ts:150]: Asking BSA

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts (hunk 4 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-claim.ts:157]: Doing this in a separate ticket

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts (hunk 5 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-claim.ts:162]: Added TODO, asking BSA whether we actually need it

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts (hunk 6 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-claim.ts:168]: Yep

File: coordinator/src/services/translation/request/dispense/dispense-claim.ts (hunk 7 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-claim.ts:244]: Yup

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts (hunk 8 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-notification.ts]: Fixed

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts (hunk 9 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-notification.ts:243]: Reduced visibility

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts (hunk 11 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-notification.ts:267]: Reduced visibility

File: coordinator/src/services/translation/request/dispense/dispense-notification.ts (hunk 13 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/request/dispense/dispense-notification.ts:360]: Yarg

File: models/fhir/common.ts (hunk 15 above)
Comment (mfjarvis) [reply models/fhir/common.ts:3]: Rich innit

File: models/fhir/common.ts (hunk 16 above)
Comment (mfjarvis) [reply models/fhir/common.ts:57]: Rich again

File: models/fhir/extension.ts (hunk 17 above)
Comment (mfjarvis) [reply models/fhir/extension.ts:28]: Double check what? Seems alright to me

File: models/hl7-v3/dispense-common.ts (hunk 18 above)
Comment (mfjarvis) [reply models/hl7-v3/dispense-common.ts]: Done some, could do more

File: coordinator/src/services/translation/common/getResourcesOfType.ts [hunk 19]
```diff
@@ -23,6 +23,14 @@ export function getMedicationDispenses(bundle: fhir.Bundle): Array<fhir.Medicati
   return getResourcesOfType<fhir.MedicationDispense>(bundle, "MedicationDispense")
 }
 
+export function getClaim(bundle: fhir.Bundle): fhir.Claim {
```
Comment (mfjarvis) [inline coordinator/src/services/translation/common/getResourcesOfType.ts:26]: No

File: coordinator/src/routes/dispense/claim.ts [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+import {fhir} from "@models"
+import * as translator from "../../services/translation/request"
+import * as claimValidator from "../../services/validation/claim-validator"
+import {spineClient} from "../../services/communication/spine-client"
+import {getScope} from "../../utils/headers"
+import {getStatusCode} from "../../utils/status-code"
+
+export default [
+  /*
+    Send a dispense release request to SPINE
```
Comment (mfjarvis) [inline coordinator/src/routes/dispense/claim.ts:18]: I don't think so

**Pull request merged**